src/syntree.rs

```diff
@@ -182,79 +285,98 @@ pub struct Expression {
 
 #[derive(Debug, Clone)]
 pub enum TypeKind {
+    /// An unspecified type that is left to the type checker.
     Implied,
+    /// A specified type by the user.
     Resolved(RuntimeType),
-    Unresolved(String),
+    /// I.e. blobs.
+    UserDefined(String),
+    /// A type that can be either `a` or `b`.
     Union(Box<Type>, Box<Type>),
+    /// `(params, return)`.
     Fn(Vec<Type>, Box<Type>),
+    /// Tuples can mix types since the length is constant.
     Tuple(Vec<Type>),
+    /// Lists only contain a single type.
     List(Box<Type>),
+    /// Sets only contain a single type.
     Set(Box<Type>),
+    /// `(key, value)`.
     Dict(Box<Type>, Box<Type>),
 }
 
+/// The constituting parts of the type system. Contains any [TypeKind].
```

Comment:
Something about it being parsed? There's a different kind of type that the type system is built on.